### PR TITLE
Add orca estimators python api docs

### PR DESCRIPTION
**Pull request by cyita**

### CODE REVIEW

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 1]
```diff
@@ -58,9 +58,25 @@ def __init__(self,
                                  batch_size=batch_size)
 
     def fit(self, data, epochs, **kwargs):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:60]: why `**kwargs`?

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 2]
```diff
@@ -58,9 +58,25 @@ def __init__(self,
...
+        """
+        Train the model with train data.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:62]: It is not supported for `OpenvinoEstimator`

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 3]
```diff
@@ -58,9 +58,25 @@ def __init__(self,
...
+
+        :param data: train data.
+        :param epochs: number of epochs to train.
+        """
         raise NotImplementedError
 
     def predict(self, data, **kwargs):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:69]: why `**kwargs`?

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 4]
```diff
@@ -58,9 +58,25 @@ def __init__(self,
...
+        """
+        Predict input data
+
+        :param data: data to be predicted. XShards and ndarray are supported. If data is XShards,
+        each element needs to be {'x': a feature numpy array or a list of feature numpy arrays}.
+        :return: predicted result.
+         If input data is XShards, the predict result is a XShards, and the schema for each result
+         is: {'prediction': predicted numpy array or list of predicted numpy arrays}.
+         If input data is numpy arrays or list of numpy arrays, the predict result is a numpy array.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:78]: need to describe the return format in more details; using examples if needed.

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 5]
```diff
@@ -134,12 +150,29 @@ def predict_transform(dict_data, batch_size):
                              "as input data, but get " + data.__class__.__name__)
 
     def evaluate(self, data, **kwargs):
+        """
+        Evaluate model.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:154]: Not supported

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 6]
```diff
@@ -134,12 +150,29 @@ def predict_transform(dict_data, batch_size):
                              "as input data, but get " + data.__class__.__name__)
 
     def evaluate(self, data, **kwargs):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:152]: why `**kwargs`?

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 7]
```diff
@@ -134,12 +150,29 @@ def predict_transform(dict_data, batch_size):
...
+        Evaluate model.
+
+        :param data: evaluation data.
+        :return: evaluation result as a dictionary of {'metric name': metric value}
+        """
         raise NotImplementedError
 
     def get_model(self):
+        """
+        Get the trained model
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:163]: Not supported

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 8]
```diff
@@ -134,12 +150,29 @@ def predict_transform(dict_data, batch_size):
...
+
+        :return: Trained model
+        """
         raise NotImplementedError
 
     def save(self, model_path):
+        """
+        Save model to model_path
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:171]: Not supported

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 9]
```diff
@@ -134,12 +150,29 @@ def predict_transform(dict_data, batch_size):
...
+
+        :param model_path: path to save the trained model.
+        :return:
+        """
         raise NotImplementedError
 
     def load(self, model_path, batch_size=0):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:178]: Need to document this method?

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 10]
```diff
@@ -168,25 +201,81 @@ def load(self, model_path, batch_size=0):
                                  batch_size=batch_size)
 
     def set_tensorboard(self, log_dir, app_name):
+        """
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:204]: Not supported; no need to document the details?

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 11]
```diff
@@ -168,25 +201,81 @@ def load(self, model_path, batch_size=0):
...
+        and validation summary (if any) will be saved to 'log_dir/app_name/validation'.
+
+        :param log_dir: The base directory path to store training and validation logs.
+        :param app_name: The name of the application.
+        """
         raise NotImplementedError
 
     def clear_gradient_clipping(self):
+        """
+        Clear gradient clipping parameters. In this case, gradient clipping will not be applied.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:219]: Not supported

File: pyzoo/zoo/orca/learn/openvino/estimator.py [hunk 12]
```diff
@@ -168,25 +201,81 @@ def load(self, model_path, batch_size=0):
...
+        In order to take effect, it needs to be called before fit.
+
+        :return:
+        """
         raise NotImplementedError
 
     def set_constant_gradient_clipping(self, min, max):
+        """
+        Set constant gradient clipping during the training process.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/openvino/estimator.py:228]: Not supported

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 13]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
                    workers_per_node=1,
                    model_dir=None,
                    backend="bigdl"):
+        """
+        Create an Estimator for torch.
+
+        :param model: PyTorch model if backend="bigdl", PyTorch model creator if backend="horovod"
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:66]: model creator function

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 14]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        or "torch_distributed"
+        :param optimizer: Orca or PyTorch optimizer if backend="bigdl", PyTorch optimizer creator
+        if backend="horovod" or "torch_distributed"
+        :param loss: PyTorch loss if backend="bigdl", PyTorch loss creator if backend="horovod"
+        or "torch_distributed"
+        :param scheduler_creator: parameter for horovod and torch_distributed. a learning rate
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:72]: parameter for "horovod" and "torch_distributed" backends

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 15]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        scheduler wrapping the optimizer. You will need to set
+        ``TorchTrainer(scheduler_step_freq="epoch")`` for the scheduler to be incremented correctly.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:74]: what is `TorchTrainer`?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 16]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        If using a scheduler for validation loss, be sure to call
+        ``trainer.update_scheduler(validation_loss)``
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:76]: what is `trainer`?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 17]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        :param training_operator_cls: parameter for horovod and torch_distributed. Custom training
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:77]: maybe no need to document this parameter as it is for internal use? @yangw1234

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 18]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        operator class that subclasses the TrainingOperator class. This class will be copied onto
+        all remote workers and used to specify custom training and validation operations.
+        Defaults to TrainingOperator.
+        :param initialization_hook: parameter for horovod and torch_distributed.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:81]: does the user need to specify this?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 19]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        :param config: parameter for horovod and torch_distributed. Config dict to create model,
+        optimizer loss and data.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:83]: does the user need to specify this?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 20]
```diff
@@ -60,6 +60,43 @@ def from_torch(*,
...
+        :param scheduler_step_freq: parameter for horovod and torch_distributed. "batch", "epoch",
+        "manual", or None. This will determine when ``scheduler.step`` is called. If "batch",
+        ``step`` will be called after every optimizer step. If "epoch", ``step`` will be called
+        after one pass of the DataLoader. If "manual", the scheduler will not be incremented
+        automatically - you are expected to call ``trainer.update_schedulers`` manually.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:88]: what is `trainer`?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 21]
```diff
@@ -173,25 +210,36 @@ def evaluate(self, data, batch_size=32, num_steps=None, profile=False, info=None
                                        profile=profile, info=info)
 
     def get_model(self):
-        """Returns the learned model(s)."""
+        """
+        Returns the learned model(s).
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:214]: Returns the learned PyTorch model.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 22]
```diff
@@ -173,25 +210,36 @@ def evaluate(self, data, batch_size=32, num_steps=None, profile=False, info=None
...
+
+        :return: The learned model(s).
+        """
         return self.estimator.get_model()
 
     def save(self, checkpoint):
-        """Saves the Estimator state to the provided checkpoint path.
+        """
+        Saves the Estimator state to the provided checkpoint path.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:222]: does it include the model?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 23]
```diff
@@ -173,25 +210,36 @@ def evaluate(self, data, batch_size=32, num_steps=None, profile=False, info=None
...
 
         :param checkpoint: (str) Path to target checkpoint file.
+        :return:
         """
         return self.estimator.save(checkpoint=checkpoint)
 
     def load(self, checkpoint):
-        """Loads the Estimator and all workers from the provided checkpoint.
+        """
+        Loads the Estimator and all workers from the provided checkpoint.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:231]: Load the Estimator state and model from from the provided checkpoint. Why workers?

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 24]
```diff
@@ -220,6 +268,25 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
 
     def fit(self, data, epochs=1, batch_size=32, feature_cols=None, labels_cols=None,
             validation_data=None, validation_metrics=None, checkpoint_trigger=None):
+        """
+        Train this torch model with train data.
+
+        :param data: train data. It can be XShards, PyTorch DataLoader and PyTorch DataLoader
+        creator.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:275]: creator function

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 25]
```diff
@@ -220,6 +268,25 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
...
+        If data is XShards, each element needs to be {'x': a feature numpy array,
+        'y': a label numpy array}
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:277]: Each partition is a dictionary of {'x': feature, 'y': label}, where feature (label) is a numpy array or a list of numpy arrays.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 26]
```diff
@@ -220,6 +268,25 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
...
+        :param epochs: Number of epochs to train the model. Default: 32.
+        :param batch_size: Batch size used for training. Only used when data is an XShards.
+        :param feature_cols: (Not supported yet) Feature column name(s) of data. Only used when data
+        is a Spark DataFrame. Default: None.
+        :param labels_cols: (Not supported yet) Label column name(s) of data. Only used when data is
+        a Spark DataFrame. Default: None.
+        :param validation_data: Validation data. SparkXShard, PyTorch DataLoader and PyTorch
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:284]: XShards

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 27]
```diff
@@ -220,6 +268,25 @@ def __init__(self, model, loss, optimizer, model_dir=None, bigdl_type="float"):
...
+        DataLoader creator are supported.
+        :param validation_metrics: Orca validation methods.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:286]: Orca metrics to be computed on validation_data

File: pyzoo/zoo/orca/learn/pytorch/estimator.py [hunk 28]
```diff
@@ -261,6 +328,17 @@ def fit(self, data, epochs=1, batch_size=32, feature_cols=None, labels_cols=None
         return self
 
     def predict(self, data, batch_size=4, feature_cols=None):
+        """
+        Predict input data.
+
+        :param data: data to be predicted. It can be XShards, each element needs to be {'x':
+        a feature numpy array}.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/pytorch/estimator.py:335]: or a list of numpy array?

File: pyzoo/zoo/orca/learn/tf/estimator.py [hunk 29]
```diff
@@ -40,39 +40,134 @@ class Estimator(SparkEstimator):
...
+        If data is XShards, each element needs to be {'x': a feature numpy array or
+        a tuple of feature numpy arrays, 'y': a label numpy array or a tuple of
+        label numpy arrays}
+        :param epochs: number of epochs to train.
+        :param batch_size: total batch size for each iteration. Default: 32.
+        :param feature_cols: feature column names if train data is Spark DataFrame.
+        :param labels_cols: label column names if train data is Spark DataFrame.
+        :param validation_data: validation data. Validation data type should be the same
+        as train data.
+        :param hard_code_batch_size: whether hard code batch size for training. Default is False.
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/tf/estimator.py:56]: when is it used?

**Comment by jason-dai** [comment]
Two quick comments:

1) If the method is not implemented, we can simply say it's not supported without documenting the details.
2) For XShards, we should describe the what data structure for each of its partition

File: pyzoo/zoo/orca/learn/openvino/estimator.py (hunk 9 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/openvino/estimator.py:171]: Already exists.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 16 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Deleted, set scheduler_step_freq="epoch" is enough.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 17 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Delete the docs of training_operator_cls=TrainingOperator and initialization_hook=None.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 20 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Deleted. the `manual` is not supported.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 22 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: It includes model, optimizer, epoch and oprator.

File: pyzoo/zoo/orca/learn/tf/estimator.py (hunk 29 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/tf/estimator.py]: Deleted since the `hard_cod_batch_size` always can be False in orca.

File: pyzoo/zoo/orca/learn/pytorch/estimator.py (hunk 15 above)
Comment (cyita) [reply pyzoo/zoo/orca/learn/pytorch/estimator.py]: Deleted, set scheduler_step_freq="epoch" is enough.

**Comment by cyita** [comment]
TODO: tf2.predict step doc

**Comment by yangw1234** [comment]
move to #3501
